### ISSUE TITLE
TreeView's MultiSelect check box is not distinguised from backplate while selected and in HighContrast.

### ISSUE THREAD

**Issue by StephenLPeters** [microsoft/microsoft-ui-xaml#2561]
Ported from internal [issue](https://microsoft.visualstudio.com/OS/_workitems/edit/26499498).

Fixing this will require some minor design guidance to decide what colors to change. 
![CheckBoxNoBorder](https://user-images.githubusercontent.com/7758786/83192348-3dedaf80-a0ea-11ea-9a50-fde7b94b6074.png)
![HC White Mode](https://user-images.githubusercontent.com/7758786/83192366-4514bd80-a0ea-11ea-95e6-c720dd81d294.png)

**Comment by StephenLPeters** [comment]
@chigy for FYI

**Comment by ranjeshj** [comment]
We should add a black or white border matching the color of the check for the checkbox.

**Comment by Felix-Dev** [comment]
Looking into this issue, I also noticed the entire checkbox "disappears" when **unselected** on pointer over (the border color has already been changed in selected state as suggested by @ranjeshj):

| Selection state | HC Black | HC Light |
|:-------------:|:-------------:|:-----:|
| Unselected      | ![treeview-highcontrast-issue](https://user-images.githubusercontent.com/1398851/83950279-d17d5980-a829-11ea-83ee-850a327a319c.gif) | ![treeview-highcontrast-issue-white](https://user-images.githubusercontent.com/1398851/83950310-17d2b880-a82a-11ea-8c3e-176ff6931f90.gif) |
| Selected     | ![treeview-highcontrast-issue-dark-selected](https://user-images.githubusercontent.com/1398851/83950527-b4498a80-a82b-11ea-8053-ab377337138c.gif) | ![treeview-highcontrast-issue-white-selected](https://user-images.githubusercontent.com/1398851/83950531-b7447b00-a82b-11ea-8ef9-3bfc369d03dc.gif) |

Perhaps we could highlight the checkbox in unselected state with a background color matching the border color when unselected? (The brush used could be the `SystemControlForegroundBaseMediumBrush` as it's the same brush used for the checkbox border when unselected.)

| HC Black | HC Light |
|:----------:|:---------:|
| ![treeview-highcontrast-issue-dark-unselected-proposal-2](https://user-images.githubusercontent.com/1398851/83951599-caa71480-a832-11ea-8697-0e37f8752a38.gif) | ![treeview-highcontrast-issue-light-unselected-proposal-2](https://user-images.githubusercontent.com/1398851/83951602-cc70d800-a832-11ea-8d8e-1181c0e54fdb.gif) |

**Comment by chingucoding** [comment]
In addition to changing the border color on hover, we should also think about changing the border color of the checkbox on an selecteditem. Having black on dark purple seems a problematic, especially if we are in high contrast which is designed to allow people with vision impairments.

**Comment by Felix-Dev** [comment]
@chingucoding Yes, I was initially thinking about that and was about to propose a checkbox border color change to white (in High Contrast Light mode) but then I thought, well, in high-contrast mode, the current color scheme will give us additional contrast between a selected item on pointer over (checkbox white border) and unselected item on pointer over (checkbox black border). This will give additional visual contrast other than the visibility state of checkbox glyph.

Perhaps we can change the checkbox border color to a ligher gray then in High Contrast Light mode (on pointer over when unselected) while keeping the white checkbox border when selected.

**Comment by StephenLPeters** [comment]
The black border on Dark purple background looks like it could be improved as well. I think the white on Purple from the selected state looks better.

**Comment by Felix-Dev** [comment]
@StephenLPeters Like this then?
![image](https://user-images.githubusercontent.com/1398851/84086499-53b37c80-a9e8-11ea-929d-4518a71da561.png)
This means the checkbox border will have the *same* color in both selected state and unselected-with-pointer-over state:
![image](https://user-images.githubusercontent.com/1398851/84086669-b278f600-a9e8-11ea-8e83-931774828236.png)

In HC Dark mode, the checkbox border will still differ depending on whether an item is selected or if it's unselected but with the pointer hovering above it:
![image](https://user-images.githubusercontent.com/1398851/84086937-519ded80-a9e9-11ea-803d-d10fd223a6f4.png)
Should we keep the checkbox border color there like it is or also align them (so change white to black border color when unselected with pointer over)? Black does seem to be easier on the eyes than white here as well.

**Comment by StephenLPeters** [comment]
I'm not a design expert but having the same border color in the selected state and unselected-with-pointer-over seems like a feature not a bug.  you can tell the difference because the selected state includes the check right?  @chigy FYI.

**Comment by Felix-Dev** [comment]
An issue arrived while trying to achieve the desired look in High-contrast mode for the Checkbox here. To recap, the desired look here is the following:
![image](https://user-images.githubusercontent.com/1398851/84819908-15861080-b019-11ea-9dc9-315bd1c59904.png)
The border color of the checkbox should be in white here when selected as it will make the checkbox far easier to see when displayed on purple background. 

The issue arrives when you take a look at the following image:
![image](https://user-images.githubusercontent.com/1398851/84827814-4cfaba00-b025-11ea-821e-886ad0ebbe2d.png)

In this image, "Root.0" is selected, "Root.1" is unselected but with the mouse hovering over it and "Root.2" is unselected and without mouse hover. As is the case when selected, the checkbox border should be white instead of black for "Root.1" to be easily visible on the purple background.

Why does this happen?
Unfortunately, outside of overriding Checkbox theme resources, I cannot easily change the border color of the checkbox (same goes for the Checkbox glyph's foreground color,....). Setting properties like Background, BorderBrush or Foreground have *no effect* on the checkbox. See the following XAML and the resulting look for example:
```xml
<CheckBox Content="Some text here"
          Foreground="Orange"
          Background="YellowGreen"
          BorderBrush="Red" BorderThickness="2"/>
```

![image](https://user-images.githubusercontent.com/1398851/84829569-43268600-b028-11ea-8194-2b8a2a28353c.png)

The only theme resources available to me here to change the color of the checkbox border are the following ones:
* CheckBoxCheckBackgroundStrokeChecked*
* CheckBoxCheckBackgroundStrokeUnchecked*

I can use these two theme resources to have a black checkbox border when the TreeViewItem is unselected and a white checkbox border when the TreeViewItem is selected. I *cannot* use these resources to change the checkbox border to white when the mouse hovers above a TreeViewItem ("Root.1" in the image above). To change the checkbox border depending on the pointer state in relation to a TreeViewItem a VisualState like TreeViewItem's `PointerOver` could be used. However, since **I cannot set the checkbox border color using properties on the checkbox control but only via overriding theme resources** I cannot just write:
```xml
<VisualState x:Name="PointerOver">
    <VisualState.Setters>
        <Setter Target="MultiSelectCheckBox.BorderBrush" Value="White" />
    </VisualState.Setters>
</VisualState>
```
That just doesn't work for the checkbox currently.

Instead, I would need to be able to override a theme resource in a visual state and I don't think that can be done elegantly, if at all.

Some possible ways to proceed with this issue comes to my mind:
1) We fix the Checkbox so that properties like `BorderBrush`, `Background` and `Foreground` can be used to change the look of the checkbox itself. This would enable me to update the checkbox border color for example in visual states of the TreeViewItem. This will probably just be a simple case of using `TemplateBindings` in the appropriate places in the Checkbox control template. This approach would also lead to the introduction of a bunch of `TreeViewItemCheckBox*` theme resources.
2) We pick a border color for the checkbox which makes it easily visible in both states: With the purple background and without it. A gray color like `#FF8C8C8C` could be used for example:
![image](https://user-images.githubusercontent.com/1398851/84831170-250e5500-b02b-11ea-894d-f4782726265a.png)
(Analog we could do the same in HC black mode if required.)
3) Keep the checkbox border color the same while unselected - no matter if the mouse is hovering above the TreeViewItem or not. The visual result is shown in this message above https://github.com/microsoft/microsoft-ui-xaml/issues/2561#issuecomment-640099473 which lead to the suggestion to change the checkbox border color further.

Approach 1) is the most powerful one as that will give us the greatest amount of control over changing the look of the checkbox depending on the visual state of the TreeViewItem it belongs to. It will probably require new code to be written in TreeViewItem.cpp and additional visual states to be added though (for example to display the correct checkbox color when the mouse is moved above the checkbox). I also recall some issues I had to work around when experimenting with #2152 because the underlying ListView is set to SelectionMode::None when the TreeView is in SelectionMode::Multiple - the mode where the Checkbox is shown. 

Approach 2) seems like a good solution to provide a fast yet acceptable fix looking to improve visibility of the checkbox in HC mode.

Approach 3) well...it would keep the changes to a minimum and only address the original issue reported.

**Comment by chigy** [comment]
@Felix-Dev , sorry for messy diagram, but is this your expectation on what you are changing about the behavior on this control? I want to make sure this is clear so that it helps triage team understand the issue you are hitting (and help me :))

![image](https://user-images.githubusercontent.com/22958680/84941244-f24d7680-b095-11ea-979d-d4e9c630e629.png)

My interpretation of the issue is that since we asked you to change the checkbox's border to be white, rest state now cannot show the border which doesn't have purple background. And we have no granular way of setting these colors per different states. Is that right?

**Comment by Felix-Dev** [comment]
@chigy 
> My interpretation of the issue is that since we asked you to change the checkbox's border to be white, rest state now cannot show the border which doesn't have purple background. 

That is correct. As your diagram shows, the checkbox border should be changed to from to white when, for example, the 
* TreeViewItem is unselected and the mouse is hovering above the item
* TreeViewItem is selected without mouse hover

In both cases, the background of the TreeViewItem is purple. 

As you correctly summarized, there is currently no easy way to set the color of the checkbox depending on the actual visual state the TreeViewItem is in. In other words, I cannot easily give the checkbox border a different color when the TreeViewItem is unselected and depending on whether the mouse may or may not hover above it. The checkbox theme resources themselves have no concept a "TreeViewItem with or without mouse hover". Setting the color of the checkbox border depending on the visual state of the TreeViewItem could be set like shown above:
```xml
<!-- TreeViewItem PointerOver visual state  -->
<VisualState x:Name="PointerOver">
    <VisualState.Setters>
        <Setter Target="MultiSelectCheckBox.BorderBrush" Value="White" />
    </VisualState.Setters>
</VisualState>
```
but this does not yet work unfortunately, as the the Checkbox currently ignores its `BorderBrush` property. As such, if the border of the checkbox is white in *TreeViewItem mouse hover* visual state (purple background) then the checkbox border color will *still* be white when the TreeViewItem is unselected and in rest mode. The checkbox border color should be black in this state though to make it visible from its white background (as depicted in your diagram).

Let me know if you still have questions or if I further need to clarify the issue I am seeing here. Hopefully though I managed to confirm your understanding of the issue at hand 🙂
